File: templates/problems/01_arrays_hashing/03_two_sum/typescript/twoSum.test.ts
import { describe, it, expect } from 'vitest';
import { twoSum } from './twoSum';

describe('TwoSum', () => {
    it('basic case', () => {
        const result = twoSum([2, 7, 11, 15], 9).sort((a, b) => a - b);
        expect(result).toEqual([0, 1]);
    });

    it('elements not adjacent', () => {
        const result = twoSum([3, 2, 4], 6).sort((a, b) => a - b);
        expect(result).toEqual([1, 2]);
    });

    it('same element value', () => {
        const result = twoSum([3, 3], 6).sort((a, b) => a - b);
        expect(result).toEqual([0, 1]);
    });

    it('negative numbers', () => {
        const result = twoSum([-1, -2, -3, -4, -5], -8).sort((a, b) => a - b);
        expect(result).toEqual([2, 4]);
    });

    it('mixed positive and negative', () => {
        const result = twoSum([-3, 4, 3, 90], 0).sort((a, b) => a - b);
        expect(result).toEqual([0, 2]);
    });

    it('large array target at end', () => {
        const result = twoSum([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 19).sort((a, b) => a - b);
        expect(result).toEqual([8, 9]);
    });
});

File: templates/problems/01_arrays_hashing/03_two_sum/go/two_sum_test.go
package two_sum

import (
	"sort"
	"testing"

	"reflect"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		name   string
		nums   []int
		target int
		want   []int
	}{
		{
			name:   "basic case",
			nums:   []int{2, 7, 11, 15},
			target: 9,
			want:   []int{0, 1},
		},
		{
			name:   "elements not adjacent",
			nums:   []int{3, 2, 4},
			target: 6,
			want:   []int{1, 2},
		},
		{
			name:   "same element value",
			nums:   []int{3, 3},
			target: 6,
			want:   []int{0, 1},
		},
		{
			name:   "negative numbers",
			nums:   []int{-1, -2, -3, -4, -5},
			target: -8,
			want:   []int{2, 4},
		},
		{
			name:   "mixed positive and negative",
			nums:   []int{-3, 4, 3, 90},
			target: 0,
			want:   []int{0, 2},
		},
		{
			name:   "large array target at end",
			nums:   []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
			target: 19,
			want:   []int{8, 9},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := twoSum(tt.nums, tt.target)
			if got == nil {
				t.Errorf("twoSum(%v, %d) = nil, want %v", tt.nums, tt.target, tt.want)
				return
			}
			// Sort both slices for order-independent comparison
			sortedGot := make([]int, len(got))
			copy(sortedGot, got)
			sort.Ints(sortedGot)

			sortedWant := make([]int, len(tt.want))
			copy(sortedWant, tt.want)
			sort.Ints(sortedWant)

			if !reflect.DeepEqual(sortedGot, sortedWant) {
				t.Errorf("twoSum(%v, %d) = %v, want %v", tt.nums, tt.target, got, tt.want)
			}
		})
	}
}
